akka-stream/src/main/scala/akka/stream/scaladsl/RetryFlow.scala

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import akka.pattern.BackoffSupervisor
+import akka.stream.scaladsl.RetryFlowCoordinator.InternalState
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler, TimerGraphStageLogic }
+import akka.stream.{ Attributes, BidiShape, FlowShape, Inlet, Outlet }
+
+import scala.collection.immutable
+import scala.concurrent.duration._
+import scala.util.Try
+
+object RetryFlow {
+
+  /**
+   * Allows retrying individual elements in the stream with an exponential backoff.
+   *
+   * The retry condition is controlled by the `retryWith` partial function. It takes an output element of the wrapped
+   * flow and should return one or more requests to be retried. For example:
+   *
+   * case Failure(_) => Some(List(..., ..., ...)) - for every failed response will issue three requests to retry
+   *
+   * case Failure(_) => Some(Nil) - every failed response will be ignored
+   *
+   * case Failure(_) => None - every failed response will be propagated downstream
+   *
+   * case Success(_) => Some(List(...)) - for every successful response a single retry will be issued
+   *
+   * A successful or failed response will be propagated downstream if it is not matched by the `retryFlow` function.
+   *
+   * If a successful response is matched and issued a retry, the response is still propagated downstream.
+   *
+   * The implementation of the RetryFlow assumes that `flow` follows one-in-one-out element semantics.
```

Comment:
What happens if it doesn't (if there is filtering)? I guess it must also not re-order elements? Doesn't that mean we are back to the issues that made us introduce WithContext in the first place (an API that doesn't allow re-ordering)?